### PR TITLE
Add TooManyRequests Exception

### PR DESCRIPTION
**Pull request by MaybeNetwork**
## Feature Summary and Justification

This adds a TooManyRequests exception.

In PRAW's [issue 1716](https://github.com/praw-dev/praw/issues/1716), 429s are being returned because Reddit [refuses to serve](https://www.reddit.com/dev/api#GET_api_morechildren) concurrent requests to the morechildren endpoint (see the note in the link), and [returns a 429](https://github.com/reddit-archive/reddit/blob/master/r2/r2/controllers/api.py#L3572) when it does get concurrent requests. I was able to get a 429 error shortly after starting two instances of PRAW that simultaneously ran

    for submission in reddit.subreddit(HIGH_TRAFFIC_SUB).top("all"):
        submission.comments.replace_more(limit= None)

I checked the ratelimit headers on the tape I made of this test and confirmed that the errors happened even with nearly all 600 requests remaining.

There are two kinds of 429 responses returned by Reddit: those that include retry-after headers, and those that don't. (Concurrent requests to the morechildren endpoint do not get retry-after headers). It wouldn't be hard to treat 429s as retry status if the response included a retry-after header, and as an exception status if the response didn't. This would be similar to how 5xx codes are treated. The rate limiter could be then adjusted to account for the presence of retry-after headers. However, since Reddit's rate limits are not always what they appear to be, this should be left for another day.

In the mean time, doing half of the above by just making 429s an exception status and adding a TooManyRequests class would not be a breaking change, since 429s already result in an assertion error.

The test `test_request__too__many_requests__without_retry_headers` was added in lieu of a test of concurrent requests to the morechildren endpoint, since the response bodies are the same.

## References

* https://www.reddit.com/dev/api#GET_api_morechildren
* https://github.com/reddit-archive/reddit/blob/master/r2/r2/controllers/api.py#L3572

### CODE REVIEW

File: prawcore/exceptions.py [hunk 1]
```diff
@@ -146,6 +146,30 @@ class TooLarge(ResponseException):
...
+        self.retry_after = response.headers.get("retry-after")
+        self.message = response.text  # Not all response bodies are valid JSON
+
+        msg = f"received {response.status_code} HTTP response"
+        msg += (
+            f". Please wait at least {float(self.retry_after)} seconds "
+            "before re-trying this request."
+            if self.retry_after
+            else ""
+        )
```
Comment (bboe) [inline prawcore/exceptions.py:169]:
```suggestion
        if self.retry_after:
            msg += f". Please wait at least {float(self.retry_after)} seconds before re-trying this request."
```

**Review by bboe — COMMENTED** [review]
Other than the one suggestion I really like this PR. Thanks!

File: prawcore/exceptions.py (hunk 1 above)
Comment (MaybeNetwork) [reply prawcore/exceptions.py:169]: I can do that. I patterned it after the message in Redirect class, which also could be changed.

**Review by bboe — APPROVED** [review]

**Comment by bboe** [comment]
Looks like this needs a quick CHANGES.rst fix since the last merge.

**Comment by bboe** [comment]
Edit, since I'm here, I'll quickly fix this one so I can do a release.

**Pull request merged**